compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp

```diff
@@ -212,4 +214,247 @@ LogicalResult setMatmulLoweringConfig(IREE::GPU::TargetAttr target,
+  // Make sure we use a tile size that results in some integral number of bytes.
+  const unsigned scaleToByte = minBitwidth < 8 ? 8 / minBitwidth : 1;
+
+  // Distribute workload to the given `numThreads` by allowing a potental loss.
+  auto distributeToThreads = [&](int64_t numThreads,
+                                 std::optional<int64_t> lossFactor =
+                                     std::nullopt) {
+    LLVM_DEBUG(llvm::dbgs() << "\nLoss factor: " << lossFactor << "\n");
+    initConfiguration();
+    // If there are more than 3 parallel dim try to tile the extra higher level
+    // dimensions to 1 for extra dimensions.
+    if (isa<linalg::GenericOp>(linalgOp.getOperation())) {
+      for (int64_t i = 0, e = workgroupTileSizes.size(); i < e; i++) {
+        if (workgroupTileSizes[i] != 0)
+          break;
+        if (loopBounds[i] != 1)
+          workgroupTileSizes[i] = 1;
+      }
+    }
+    // Scan from the innermost shape dimension and try to deduce the
+    // configuration for the corresponding GPU workgroup dimension.
+    int64_t wgDim = 0;
+    for (auto shapeDim : llvm::reverse(partitionableLoops)) {
+      int64_t loopBound = loopBounds[shapeDim];
+      // Skip dynamic dimensions.
+      if (ShapedType::isDynamic(loopBound))
+        continue;
+
+      // Try to find some power of two that can devide the current shape dim
+      // size. This vector keeps the candidate tile sizes.
+      SmallVector<int64_t, 8> candidates;
+
+      // For the inner most workgroup dim, try to see if we can have 4
+      // elements per thread. This enables vectorization.
+      if (vectorizable && wgDim == 0 && !lossFactor) {
+        candidates.push_back(4 * numThreads);
+      }
+      // Try all power of two numbers upto the subgroup size.
+      for (unsigned i = numThreads; i >= 1; i >>= 1) {
+        candidates.push_back(i);
+      }
+      LLVM_DEBUG({
+        llvm::dbgs() << "Base candidate tile sizes: [";
+        llvm::interleaveComma(candidates, llvm::dbgs());
+        llvm::dbgs() << "]\n";
+      });
+
+      for (int64_t candidate : candidates) {
+        int64_t scaledTileSize = candidate * scaleToByte;
+        if (loopBound % scaledTileSize != 0) {
+          if (!lossFactor)
+            continue;
+          // Skip this candidate if it causes many threads to be idle.
+          int64_t idleThreads = candidate - (loopBound % scaledTileSize);
+          if (idleThreads > candidate / *lossFactor)
+            continue;
+        }
+        // If the workload is too small and we cannot distribute to more than 2
+        // workgroups, try a smaller tile size to increase parallelism.
+        if (partitionableLoops.size() == 1 && candidate > subgroupSize &&
+            llvm::divideCeil(loopBound, scaledTileSize) <= 2) {
+          continue;
+        }
+
+        // Found a suitable candidate. Try to let each thread handle 4
+        // elements if this is the workgroup x dimension.
```

Comment:
Shouldn't we configure for 4 dwords worth of input elements?

Reply:
Probably, but this is supposed to be a fallback path that we can start to replace some older pipelines with. Right now I am just importing the well tested logic on the SPIR-V side. Also dwordx4 loads runs the risk of large vector sizes for smaller bit widths so I'd rather keep this a simple copy + paste until we are using this for real and have actual real world examples to look at for it.

Reply:
I am going to land this for now and we can revisit potential improvements like this in the future.